src/main/scala/mesosphere/marathon/api/akkahttp/v2/AppsController.scala

```diff
@@ -6,90 +6,131 @@ import java.time.Clock
+import mesosphere.marathon.core.task.tracker.InstanceTracker
+import mesosphere.marathon.core.task.tracker.InstanceTracker.InstancesBySpec
+import mesosphere.marathon.core.task.Task.{ Id => TaskId }
 
 import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success }
+import scala.util.control.NonFatal
 
 class AppsController(
     val clock: Clock,
     val eventBus: EventStream,
-    val appTasksRes: AppTasksResource,
-    val service: MarathonSchedulerService,
+    val marathonSchedulerService: MarathonSchedulerService,
     val appInfoService: AppInfoService,
+    val healthCheckManager: HealthCheckManager,
+    val instanceTracker: InstanceTracker,
+    val taskKiller: TaskKiller,
     val config: MarathonConf,
     val groupManager: GroupManager,
     val pluginManager: PluginManager)(
     implicit
     val actorSystem: ActorSystem,
+    val materializer: Materializer,
     val executionContext: ExecutionContext,
     val authenticator: MarathonAuthenticator,
     val authorizer: Authorizer,
     val electionService: ElectionService
 ) extends Controller {
   import Directives._
+  import RichEither.RightBiased
 
   private implicit lazy val validateApp = AppDefinition.validAppDefinition(config.availableFeatures)(pluginManager)
   private implicit lazy val updateValidator = AppValidation.validateCanonicalAppUpdateAPI(config.availableFeatures, () => normalizationConfig.defaultNetworkName)
 
-  import AppsController._
+  import AppHelpers._
   import EntityMarshallers._
 
   import mesosphere.marathon.api.v2.json.Formats._
 
+  private val forceParameter = parameter('force.as[Boolean].?(false))
+
   private def listApps(implicit identity: Identity): Route = {
     parameters('cmd.?, 'id.?, 'label.?, 'embed.*) { (cmd, id, label, embed) =>
       def index: Future[Seq[AppInfo]] = {
-        def containCaseInsensitive(a: String, b: String): Boolean = b.toLowerCase contains a.toLowerCase
-
-        val selectors = Seq[Option[Selector[AppDefinition]]](
-          cmd.map(c => Selector(_.cmd.exists(containCaseInsensitive(c, _)))),
-          id.map(s => Selector(app => containCaseInsensitive(s, app.id.toString))),
-          label.map(new LabelSelectorParsers().parsed),
-          Some(authzSelector)
-        ).flatten
         val resolvedEmbed = InfoEmbedResolver.resolveApp(embed.toSet) + AppInfo.Embed.Counts + AppInfo.Embed.Deployments
-        appInfoService.selectAppsBy(Selector.forall(selectors), resolvedEmbed)
+        val selector = selectAuthorized(search(cmd, id, label))
+        appInfoService.selectAppsBy(selector, resolvedEmbed)
       }
       onSuccess(index)(apps => complete(Json.obj("apps" -> apps)))
     }
   }
 
-  private def createApp(app: AppDefinition, force: Boolean)(implicit identity: Identity): Route = {
-    def create: Future[(DeploymentPlan, AppInfo)] = {
+  private[v2] def search(cmd: Option[String], id: Option[String], label: Option[String]): AppSelector = {
```

Comment:
Is this method tested? Could you add docs? What is a selector?

Reply:
It is :) it is called indirectly when we pass a http request to the route.